encoding/ssz/query/generalized_index.go

```diff
@@ -0,0 +1,358 @@
+					}
+					multiplier = nextPowerOfTwo(innerChunkCount)
+					offset = *idx
+				}
+				root = updateRoot(root, 1, multiplier, offset)
+				currentInfo = elem
+
+			case Bitlist:
+				// Bits packed into 256-bit chunks; select the chunk containing the bit
+				chunkPos := *idx / bitsPerChunk
+				innerChunkCount, err := getChunkCount(fieldSsz)
+				if err != nil {
+					return 0, fmt.Errorf("chunk count error: %w", err)
+				}
+				root = updateRoot(root, listBaseIndex, innerChunkCount, chunkPos)
+				// Bits element is not further descendable; set to basic to guard further steps
+				currentInfo = &sszInfo{sszType: Boolean, fixedSize: 1}
+
+			case Bitvector:
+				chunkPos := *idx / bitsPerChunk
+				innerChunkCount, err := getChunkCount(fieldSsz)
+				if err != nil {
+					return 0, fmt.Errorf("chunk count error: %w", err)
+				}
+				root = updateRoot(root, 1, innerChunkCount, chunkPos)
+				// Bits element is not further descendable; set to basic to guard further steps
+				currentInfo = &sszInfo{sszType: Boolean, fixedSize: 1}
+
+			default:
+				return 0, fmt.Errorf("indexing not supported for type %s", fieldSsz.sszType)
+			}
+			continue
+		}
+	}
+
+	return root, nil
+}
+
+// updateRoot computes the new generalized index based on the current root, base index, chunk count, and offset
+// base index is typically 1 for containers and 2 for lists
+// root = root * base_index * pow2ceil(chunk_count(container)) + fieldPos
+func updateRoot(root uint64, baseIndex uint64, chunkCount uint64, offset uint64) uint64 {
+	return root*baseIndex*nextPowerOfTwo(chunkCount) + offset
+}
+
+// isBasicType checks if the SSZType is a basic type
+func isBasicType(sszType SSZType) bool {
+	switch sszType {
+	case UintN, Byte, Boolean:
+		return true
+	default:
+		return false
+	}
+}
+
+// getChunkCount returns the number of chunks for the given SSZInfo (equivalent to chunk_count in the spec)
+func getChunkCount(info *sszInfo) (uint64, error) {
+	switch info.sszType {
+	case UintN, Byte, Boolean:
+		return 1, nil
+	case Container:
+		containerInfo, err := info.ContainerInfo()
+		if err != nil {
+			return 0, err
+		}
+		return uint64(len(containerInfo.order)), nil
+	case List:
+		listInfo, err := info.ListInfo()
+		if err != nil {
+			return 0, err
+		}
+		// For Lists with basic element types, multiple elements can be packed into 32-byte chunks
+		elementInfo, err := listInfo.Element()
+		if err != nil {
+			return 0, err
+		}
+		elemLength := itemLengthFromInfo(elementInfo)
+		return (listInfo.Limit()*uint64(elemLength) + 31) / bytesPerChunk, nil
+	case Vector:
+		vectorInfo, err := info.VectorInfo()
+		if err != nil {
+			return 0, err
+		}
+		// For Vectors with basic element types, multiple elements can be packed into 32-byte chunks
+		elementInfo, err := vectorInfo.Element()
+		if err != nil {
+			return 0, err
+		}
+		elemLength := itemLengthFromInfo(elementInfo)
+		return (vectorInfo.Length()*uint64(elemLength) + 31) / bytesPerChunk, nil
```

Comment:
```suggestion
		return (vectorInfo.Length()*elemLength + 31) / bytesPerChunk, nil
```

Other nit, we don't need type cast here as it is already `uint64`. Same comment for `List` case above.